R/geo.R

```diff
@@ -0,0 +1,217 @@
+        halt("None of the geographies match at all. Either the variable is",
+             " wrong, or Crunch doesn't yet have geodata for this variable.")
+    }
+    if (nrow(match_scores) > 1) {
+        halt("There is more than one possible match. Please specify the geography manually:\n",
+             paste0(capture.output(print(match_scores[,c("value", "geodatum_name", "geodatum")])), collapse = "\n"))
+    }
+    
+    match_geo <- new('CrunchGeography', 
+                     list(geodatum=match_scores$geodatum,
+                          feature_key=paste0('properties.',match_scores$property),
+                          match_field=match_field))
+    return(match_geo)
+}
+
+#' @rdname crunch-is
+#' @export
+is.Geodata <- function (x) inherits(x, "Geodata")
+
+#' @rdname geo
+#' @export
+availableGeodata <- function (x = getAPIRoot()) {
+    return(GeoCatalog(crGET(shojiURL(x, "catalogs", "geodata"))))
+}
+
+#' all available property features for all available geographies
+#' 
+#' @param x an API root address (default: the R-session default)
+#' @param geodatum_fields character, what peices of information about each 
+#' geodatum should be retained? (default: `c("name", "description", "location")``)
+#' 
+#' @return a (tidy) dataframe with all of the available features and 
+#' geographies for matching
+#' 
+#' @export
+availableFeatures <- function (x = getAPIRoot(), geodatum_fields=c("name", "description", "location")) {
+    geo_cat <- availableGeodata(x)
+    
+    # grab each geodatum in order to get metadata
+    # TODO: this should probably not be done through a series of GETs, 
+    # but the current API requires it.
+    geo_metadatas <- lapply(urls(geo_cat), function(x) Geodata(crGET(x)) )
+    names(geo_metadatas) <- urls(geo_cat)
+    
+    out <- lapply(geo_metadatas, function (geography) {
+        meta <- geography$metadata$properties
+        lapply(names(meta), function (x) {
+            # remove nulls
+            values <- unlist(meta[[x]])
+            values[is.null(values)] <-  NA_character_
+            data.frame(property=x, value=values)
+        })
+    })
+    
+    out <- lapply(names(out), function (x) {
+        dfs <- do.call("rbind", out[[x]])
+        dfs$geodatum <- x
+        dfs[paste0("geodatum_", geodatum_fields)] <- geo_cat[[x]][geodatum_fields]
+        return(dfs)
+    })
+    return(do.call("rbind", out))
+}
+
+
+#' Score similarity between a feature dataframe and categories
+#' 
+#' Implemented using the Jaccard index, where a number closer to 1 is 
+#' more similar.
+#' 
+#' @param features a vector of features to match (usually from a subset of the
+#' output `[availableFeatures]`) with a single property for a single geodatum.
+#' @param categories a vector of categories to match
+#' 
+#' @return the Jaccard index for the values of the property given in 
+#' feat_df and the vector of categories
+#' 
+#' @export
+scoreCat2Feat <- function (features, categories) {
```

Comment:
The norm in the package currently is to use "To" instead of "2" in functions that convert (like `toVariable`, `catalogToDataFrame`, etc.)